Atualmente se passarmos um ID que não existe na procura de um médico, receberemos um erro do tipo `500 Internal Server Error` e, em "Preview", temos uma mensagem com os campos `timestamp`, `status`, `error` e `trace`.

``` Json
{
    "timestamp": "2022-10-21T17:59:29.080+00:00",
    "status": 500,
    "error": "Internal Server Error",
    "trace":
"jakarta.persistence.EntityNotFoundException: Unable to find med.voll.api.medico.Medico with id 6999\n\tatorg.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderimpl$JpanEntityNoFoundDelegate"

//retorno omitido
}
```

`Código ou erro 500: Erro no servidor interno`.

Nas informações do `JSON` há um objeto com algumas propriedades. A primeira é o `timestamp` que comunica data e hora do erro, o `status` que é o código HTTP `500` e o `error` é o nome do erro referente ao código.

Por fim, a propriedade `trace`, sendo stack trace do erro - sendo a mesma exibida no console, informando qual a exception que ocorreu, nesta mensagem exibe que o ID que digitamos não existe no banco de dados.

Esse é o JSON devolvido. Porém, não é uma boa prática retornarmos um stack trace para o usuário da API, seja em uma aplicação front-end ou aplicativo mobile que está consumindo a nossa API.

Isso porque estamos expondo informações sensíveis e desnecessárias, o que pode  se tornar uma brecha de segurança.

Neste caso da requisição `get`, poderíamos informar somente o status e qual o erro. Vamos aprender a fazer esse tratamento no retorno.

Desejamos padronizar as mensagens devolvidas pela API quando ocorre erro. Há diversas situações de erro que podem ocorrer da nossa API, nem sempre tudo funcionará conforme o esperado.

Por isso, em situações de erros é uma boa prática colocar respostas mais apropriadas para esses cenários.

Precisamos perceber que o próprio tratamento do Spring já traz informações que desejamos, porém, a propriedade `trace` nos devolve dados desnecessários e sensíveis.

Então o JSON devolvido no Insomnia é o padrão do Spring Boot, mas podemos configurar no projeto para não retornar os dados do campo `trace`.

## Modificando retorno de erro
No arquivo `application.properties` em src -> main -> resources, temos as propriedades que configuramos de banco de dados e de gerar o SQL toda vez que a API vai no banco de dados. Agora, incluiremos mais uma propriedade para o Spring não enviar a *stack trace* em caso de erro.

Na documentação oficial do Spring, lá temos um menu com o agrupamento das propriedades conforme a categoria. Em `Server Properties`, temos uma tabela com os campos: *name*, *description* e *default value*. Nela, temos as propriedades ideal para o nosso caso, sendo a `server.error.include-stacktrace`.

Então vamos colocar essa propriedade no `application.properties` com o valor `never`, na última linha do arquivo.

``` Java
spring.datasource.url=jdbc:postgres://localhost:5432/vollmed
spring.datasource.username=postgres
spring.datasource.password=postgres

spring.jpa.show-sql=true
spring.jpa.properties.hibernate.format_sql=true

server.error.include-stacktrace=never
```

Assim, conseguimos aplicar um tratamento mais apropriado para esse cenário. Caso dê erro 500 na API, devolvemos um JSON (sendo que o próprio Spring fez isso, não escrevemos nenhum código) e somente informamos para não ser exibido o stack trace.

O usuário da API precisa saber somente que ocorreu um erro, qual o erro, o endereço que ocorreu e uma mensagem informando o motivo.

Mas na verdade, esta situação não deveria gerar um código `500`. Se estamos disparando uma requisição com um número de ID inexistente, o código HTTP mais adequando é o `404`.

`Código ou erro 404: Recurso não encontrado`.

O código `500` ocorreu porque estamos usando o Spring Data, e fazendo uma consulta no banco de dados usando a interface repository. E o padrão do Spring Data JPA ao passar um ID inexistente é o retornar uma exception. Assim, ao gerar uma exception, nos retorna o código ou erro `500`.

Podemos configurar isso no nosso projeto. Em casos de determinadas exceptions, não retornar erro 500 e sim o 404. Vamos personalizar o tratamento de erros na API.

## Tratando erro 404
Para ajustarmos o retorno do erro 404 em vez do erro 500, vamos no arquivo `MedicoController`. O método que está sendo chamado nessa requisição é o `detalhar`.

Neste método, ele chama o `repository.getReferenceById(id)`. É justamente nesta linha que está o problema. O método `getReferenceById()`, ao passarmos um ID inexistente na tabela do banco de dados, ele envia uma exception do tipo `EntityNotFoundException`.

E se percebermos não fizemos o tratamento, não temos o `try-catch`. Assim, a exception aconteceu nessa linha e foi lançada para o Spring. O Spring, por padrão, se ocorrer uma exception que não foi tratada no código, ele trata como uma exceção gerando um erro `500`.

`Por padrão, exceções não tratadas no código são interpretadas pelo Spring Boot como erro 500`.

Não desejamos isso, desejamos o erro `404`. Uma forma de lidar com essa situação, seria isolar o código do métodod de detalhar, dentro de um `try-catch`. Faremos a captura da exception, e devolveremos o código `404`.

Essa exception pode acontecer em outros métodos ou controllers, então temos que tratar melhor que somente em método por método com um `try-catch`. Então vamos usar outro recurso do Spring para isolar esse tipo de tratamento de erros.

Vamos criar uma classe e nela termos o método responsável por tratar esse erro em específico.

Vamos criar um pacote chamado "infra", em que ficarão os códigos relacionados à infraestrutura.

E criamos a classe `TratadorDeErros`.

Por enquanto, é uma classe em Java e não há nada em Spring. Isto é, o Spring não irá carregar essa classe automaticamente quando recarregarmos o projeto. Para ele carregar essa classe, é o necessário termos à anotação `@RestControllerAdvice`.

`TratadorDeErros`:
``` Java
package med.voll.api.infra;

import org.springframework.web.bind.annotation.RestControllerAdvice;

@RestControllerAdvice
public class TratadorDeErros {

}
```

Nesta classe, criaremos um método responsável por lidar com a exceção `EntityNotFoundException`. vamos criar o método `tratarErro404()`.

``` Java
@RestControllerAdvice
public class TratadorDeErros {

    public void tratarErro404() {

    }
}
```

Uma linha anterior ao método, precisamos informar ao Spring para *qual exceção esse método será chamado*. No caso, usaremos à anotação `@ExceptionHandler` e, dentro de um parênteses, vamos especificar a classe exception:

``` Java
@ExceptionHandler(EntityNotFoundException.class)
```

``` Java
@RestControllerAdvice
public class TratadorDeErros {

    @ExceptionHandler(EntityNotFoundException.class)
    public void tratarErro404() {

    }
}
```

Desta maneira, com essa anotação, o Spring sabe que se em qualquer controller do projeto for lançado uma exceção `EntityNotFoundException`, é para chamar o método `tratarErro404()`. E o que devolvermos, é o que será devolvido como resposta na requisição.

Então para isso, precisamos alterar de `void` para `ResponseEntity`.

``` Java
@ExceptionHandler(EntityNotFoundException.class)
public ResponseEntity tratarErro404() {

}
```

No método `tratarErro404()`, colocaremos um `return ResponseEntity.notFound()`. No final, incluiremos um `.build()` para ele criar o objeto `Response Entity`.

``` Java
@ExceptionHandler(EntityNotFoundException.class)
public ResponseEntity tratarErro404() {
    return ResponseEntity.notFound().build();
}
```

Então se percebermos é simples criar uma classe com métodos que trataram exceções não tratadas no controller. Logo, esse código fica isolado e não precisamos ter `try-catch` nos controllers. Estes nem precebem que há uma classe externa tratando o erro.

Agora se procurarmos um médico com o ID que não existe, vamos receber o código `404 Not Found`.

## Tratar erro 400
Este código ocorre quando o cliente da nossa API dispara uma requisição com dados inválidos.

`Código ou erro 400: indica que o servidor não conseguiu processar uma requisição por erro de validação nos dados enviados pelo cliente`.

Estamos usando o *bean validation* na API para realiar as validações dos campos obrigatórios, entre outros dados.

Então inserindo os tais dados inválidos, no momento estamos recebendo o erro `400 Bad Request`. Por padrão, ao acontecer um erro `400`, o Spring faz a integração com o *bean validation*, e roda as validações.

Caso ocorra alguma falha, é devolvido o código 400 e no JSON da resposta nos envia vários campos, como timestamp, status, error, message, entre outros campos. Há, também, um objeto chamado errors, que contém o array com cada erro de validação ocorrido.

No entanto, esse array possui um objeto complexo com várias informações.

Como um exemplo de erro:

``` JSON
// Retorno omitido
{
    "codes": [
        "NotBlank.dadosCadastroMedico.telefone",
        "NotBlank.telefone",
        "NotBlank.java.lang.string",
        "NotBlank"
],

// Retorno omitido
}
```

temos os nomes das anotações do *bean validation*. Contudo, esse retorno poderia ser mais simplificado.

Podemos retornar um JSON com uma lista dos campos que geraram erro e, para cada campo, devolver qual o campo e a mensagem de erro. Por exemplo: "campo nome é obrigatório", "o email está com formato inválido", etc.

Então na classe `TratadorDeErros` vamos adicionar um novo método. Criaremos mais um método chamado `tratarErro400()`.

```Java
public ResponseEntity tratarErro400() {

}
```

Agora precisamos da anotação em cima do método. Qual a exception que é lançada pelo Spring ou *bean validation*?

Há uma específica para esse erro, do *bean validation*, o `MethodArgumentNoValidException.class`.

``` Java
@ExceptionHandler(MethodArgumentNotValidExcption.class)
public ResponseEntity tratarErro400() {

}
```

Essa é a exceção que o *bean validation* lança quando há campo inválido. Assim, o Spring já sabe que se em alguma requisição ou controller ocorrer uma exception do tipo `Method Argument Not Valid Exception`, é para cair no erro `400`.

No retorno, colocaremos `return ResponseEntity.badRequest().build()`.

``` Java
@ExceptionHandler(MethodArgumentNotValidException.class)
public ResponseEntity tratarErro400() {
    return ResponseEntity.badRequest().build();
}
```

Agora se tentarmos adicionar um médico faltando com dados inválidos, nos é devolvido o código `400 Bad Request`. Mas na preview não retornou nenhum dado, somente a mensagem "No body returned for response". Conseguimos fazer o tratamento personalizado, mas faltou o corpo da resposta.

Para retornarmos um corpo com as informações no método `tratarErro400()`, precisamos passar o objeto que desejamos retornar no parênteses do `bad request()`.

Porém, agora diferente do código `404`, no código `400` é necessário sabermos quais erros ocorreram. Portanto, precisamos capturar a exceção lançada, pois é nela que teremos acesso a quais campos estão inválidos conforme as regras do bean validation.

Para isso, na assinatura do método `tratarErro400()`, receberemos como parâmetro a `exception` lançada. Essa exceção deve ser a mesma da anotação, podemos copiá-la.

Logo após, colocaremos no parênteses do método `tratarErro400()`.

```Java
@ExceptionHandler(MethodArgumentNotValidException.class)
public ResponseEntity tratarErro400(MethodArgumentNotValidException ex) {
    return ResponseEntity.badRequest().build();
}
```

Se quisermos, prodemos receber a *exception* no método que estamos tratando o erro, basta declarar como parâmetro na assinatura do método. Para capturar os erros que ocorreram no *bean validation*.

Esse objeto exception possui um método que retorna a lista com os campos inválidos. Então colocamos isso em uma variável chamada `erros` dentro do método, antes do `return`.

``` Java
@ExceptionHandler(MethodArgumentNotValidException.class)
public ResponseEntity tratarErro400(MethodArgumentNotValidException ex) {
    var erros = ex.getFieldErrors();
    return ResponseEntity.badRequest().build();
}
```

O `get field errors()` lista cada erro ocorrido em cada campo. E esse é o objeto que desejamos retornar no corpo da resposta.

Guardamos o `getFieldErros` na variável `erros`, e para enviar um corpo na resposta incluímos o `.body()` no return. No método `body()` passamos a lista de `erros`, e podemos remover o `.build()`, já que o body nos retorna o objeto `ResponseEntity`.

``` Java
@ExceptionHandler(MethodArgumentNotValidException.class)
public ResponseEntity tratarErro400(MethodArgumentNotValidException ex) {
    var erros = ex.getFieldErrors();

    return ResponseEntity.badRequest().body(erros);
}
```

Porem do jeito que está ele nos retorna na Preview o mesmo retorno anterior, com informações desnecessárias. No caso, desejamos retornar somente o nome do campo e a mensagem.

Para isso, vamos personarlizar aquela lista, transformando-a em um DTO. Assim como fizemos na lista de médicos, precisaremos criar um DTO que contém somente os campos e as mensagens. E teremos que converter a lista do bean validation, para a lista do nosso DTO.

```
DTO: data transfer object
```

Então vamos criar a DTO, mas criaremos dentro da classe `TratadorDeErros`. Isso porque vamos usar o DTO somente dentro dessa classe. Logo, podemos criar como um record interno. Vamos declarar o record chamando de `DadosErroValidacao()`.

``` Java
private record DadosErroValidacao() {
    
}
```

Esse DTO terá somente dois parâmetros: campo e mensagem.

``` Java
private record DadosErroValidacao(String campo, String mensagem) {

}
```

Agora, no `body` do método `tratarErro400` não passaremos mais a lista de `erros`, precisamos convertê-la para uma lista de dados, erro e validação.

Então para converter em uma lista de dados, vamos utilizar o método `map`. Passaremos `erros.stream().map()`. Essa parte solicita para os erros que me dê um stream e mapeie cada objeto `FieldError` para um objeto `DadosErroValidacao`.

Em `DadosErroValidacao` acrescentamos `::new` para chamar o construtor. Por fim, usaremos o método `toList()` para convertermos para uma lista.

``` Java
@ExceptionHandler(MethodArgumentNotValidException.class)
public ReponseEntity tratarErro400(MethodArgumentNotValidException ex) {
    var erros = ex.getFieldErrors();

    return ResponseEntity.badRequest().body(erros.stream().map(DadosErroValidacao::new).tolist());
}
```

Precisamos incluir no nosso DTO `DadosErroValidacao` outro construtor que receberá o objeto `FieldError`.

``` Java
private record DadosErroValidacao(String campo, String mensagem) {
    public DadosErroValidacao(FieldError erro) {

    }
}
```

Então agora, ainda no `record`, chamaremos o construtor padrão do `record` passando `erro.getField()` e `erro.getDefaultMessage()`.

- **erro.getField()** : nos devolve o nome do campo
- **erro.getDefaultMessage()** : nos devolve a mensgaem para um campo específico.

``` Java
private record DadosErroValidacao(String campo, String mensagem) {
    public DadosErroValidacao(FieldError erro) {
        this(erro.getField(), erro.getDefaultMessage());
    }
}
```

Desta maneira agora, quando inserimos os dados inválidos o Insomnia mostra qual os campos errados.

Exemplo: 

``` JSON
[
    {
    "campo": "telefone",
    "mensagens": "must not be blank"
    },
    {
    "campo": "email",
    "mensagens": "must not be blank"
    },
        {
    "campo": "nome",
    "mensagens": "must not be blank"
    },
        {
    "campo": "crm",
    "mensagens": "must not be blank"
    }
]
```
Temos um array com cada campo inválido e sua respectiva mensagem.